libs/analysis/aframe/analysis/sensitivity.py

```diff
@@ -102,68 +119,91 @@ def __post_init__(self):
+        self, all_parameters, target: "bilby.core.prior.PriorDict"
+    ):
+        source_only = [key for key in self.source if key not in target]
+        target_only = [key for key in target if key not in self.source]
+
+        # First, remove any keys that are in only the source prior
+        source = self.source
+        for key in source_only:
+            source.pop(key)
+
+        # If the source prior doesn't have a distribution that's in
+        # the target prior, approximate a distribution based on the
+        # sampled events so that we can call .prob()
+        for key in target_only:
+            values = [sample[key] for sample in all_parameters]
+            grid, pdf = pdf_from_events(values)
+            source[key] = Interped(grid, pdf, np.min(grid), np.max(grid))
+
+        return source
+
     def weights(
         self,
-        recovered_parameters: Dict[str, np.ndarray],
+        samples: List[Dict[str, float]],
+        source: "bilby.core.prior.PriorDict",
         target: "bilby.core.prior.PriorDict",
     ):
         """
         Calculate the weights for the samples.
         """
         weights = []
-        for sample in recovered_parameters:
-            weight = target.prob(sample) / self.source.prob(sample)
+        for sample in samples:
+            weight = target.prob(sample) / source.prob(sample)
             weights.append(weight)
         return np.array(weights)
 
     def __call__(
         self,
         recovered_parameters: List[Dict[str, float]],
-        rejected_parameters: List[Dict[str, float]],
-        num_injections: int,
+        all_parameters: List[Dict[str, float]],
         target: Optional["bilby.core.prior.PriorDict"] = None,
     ):
         """
         Calculate the sensitive volume and its uncertainty.
 
         Args:
             recovered_parameters:
-                List of recovered parameters. All quantities should be in
-                the source frame.
-            num_injections:
-                Number of total injections. This includes any injections
-                not used during rejection sampling.
+                List of samples that produced recovered injections
+            all_parameters:
+                List of all parameters that were generated, including
+                any that were rejected by an SNR cut
             target:
                 Bilby PriorDict of the target distribution, to which the
                 samples will be reweighted.
                 If None, the target is assumed to be the source distribution.
         """
+
+        # Equation 9 of https://arxiv.org/pdf/1712.00482.pdf
+        # If our injections were sampled in chirp distance, our volume
+        # needs to be modified
+        if "chirp_distance" in self.source and target is not None:
+            num_samples = 100000
+            target_samples = target.sample(num_samples)
+            num_inside = sum(self.source.prob(target_samples) > 0)
```

Comment:
You have to add `axis=0` if you want per-sample estimates, otherwise it will aggregate the probability over all samples